platforms/core-configuration/file-collections/src/main/java/org/gradle/api/internal/file/collections/ReproducibleDirectoryWalker.java

```diff
@@ -85,5 +84,35 @@ private static FileVisitResult visit(Path path, PathVisitor pathVisitor) throws
         }
     }
 
-    private final static Comparator<Path> FILES_FIRST = Comparator.<Path, Boolean>comparing(x -> x.toFile().isDirectory()).thenComparing(Path::toString);
+    private static PathWithAttributes toPathWithAttributes(Path path) {
+        try {
+            BasicFileAttributes attributes = Files.readAttributes(path, BasicFileAttributes.class);
+            return new PathWithAttributes(path, attributes, null);
+        } catch (IOException ignored) {
+            try {
+                BasicFileAttributes attributes = Files.readAttributes(path, BasicFileAttributes.class, LinkOption.NOFOLLOW_LINKS);
+                return new PathWithAttributes(path, attributes, null);
+            } catch (IOException e) {
+                return new PathWithAttributes(path, null, e);
+            }
+        }
+    }
+
+    private static class PathWithAttributes {
+        private final Path path;
+        @Nullable
+        private final BasicFileAttributes attributes;
+        @Nullable
+        private final IOException exception;
+
+        public PathWithAttributes(Path path, @Nullable BasicFileAttributes attrs, @Nullable IOException exception) {
+            this.path = path;
+            this.attributes = attrs;
+            this.exception = exception;
+        }
+
+        public boolean isDirectory() {
+            return attributes != null && attributes.isDirectory();
+        }
+    }
```

Comment:
💅 Can we use records for these things now?

Reply:
Unfortunately not yet, since we don't set language level to 17 yet on latest master